Python/legacy_tracing.c

```diff
@@ -500,87 +535,155 @@ _PyEval_SetProfile(PyThreadState *tstate, Py_tracefunc func, PyObject *arg)
         return -1;
     }
 
-    // needs to be decref'd outside of the lock
-    PyObject *old_profileobj;
-    FT_MUTEX_LOCK(&_PyRuntime.ceval.sys_trace_profile_mutex);
-    Py_ssize_t profiling_threads = setup_profile(tstate, func, arg, &old_profileobj);
-    FT_MUTEX_UNLOCK(&_PyRuntime.ceval.sys_trace_profile_mutex);
-    Py_XDECREF(old_profileobj);
+    PyInterpreterState *interp = tstate->interp;
+    if (_PyOnceFlag_CallOnce(&interp->sys_profile_once_flag,
+                             setup_profile_callbacks, NULL) < 0) {
+        return -1;
+    }
 
-    uint32_t events = 0;
-    if (profiling_threads) {
-        events =
-            (1 << PY_MONITORING_EVENT_PY_START) | (1 << PY_MONITORING_EVENT_PY_RESUME) |
-            (1 << PY_MONITORING_EVENT_PY_RETURN) | (1 << PY_MONITORING_EVENT_PY_YIELD) |
-            (1 << PY_MONITORING_EVENT_CALL) | (1 << PY_MONITORING_EVENT_PY_UNWIND) |
-            (1 << PY_MONITORING_EVENT_PY_THROW);
+    _PyEval_StopTheWorld(interp);
+    PyObject *old_profileobj = swap_profile_func_arg(tstate, func, arg);
+    int ret = set_monitoring_profile_events(interp);
+    _PyEval_StartTheWorld(interp);
+    Py_XDECREF(old_profileobj);  // needs to be decref'd outside of stop-the-world
+    return ret;
+}
+
+int
+_PyEval_SetProfileAllThreads(PyInterpreterState *interp, Py_tracefunc func, PyObject *arg)
+{
+    PyThreadState *current_tstate = _PyThreadState_GET();
+    assert(is_tstate_valid(current_tstate));
+    assert(current_tstate->interp == interp);
+
+    if (_PySys_Audit(current_tstate, "sys.setprofile", NULL) < 0) {
```

Comment:
Shouldn't this be `sys._setprofileallthreads`?

Reply:
I think it's less risky to keep the auditing even name the same as it was before.